src/Raven.Client/ServerWide/Tcp/TcpNegotiation.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Client.ServerWide.Tcp
+{
+    public class TcpNegotiation
+    {
+        public static int NegotiateProtocolVersion(JsonOperationContext documentsContext, Stream stream, TcpNegotiateParamaters parameters)
+        {
+            using (var writer = new BlittableJsonTextWriter(documentsContext, stream))
+            {
+                var currentVersion = parameters.Version;
+                while (true)
+                {
+                    documentsContext.Write(writer, new DynamicJsonValue
```

Comment:
I think it would be better to 'invert' the responsibility and send list of all supported protocol versions for this particular operations so destination can make a decision and send back the version that should be used.